dataset.go

```diff
@@ -86,6 +114,70 @@ func flatElementsIterator(elems []*Element, elemChan chan<- *Element) {
 	}
 }
 
+// FlatDatasetIterator represents a simple iterator over the elements in a
+// Dataset.
+type FlatDatasetIterator interface {
+	// HasNext indicates if the iterator as another element.
+	HasNext() bool
+	// Next gets and returns the next element in the iterator.
+	Next() *Element
+}
+
+// FlatStatefulIterator returns a stateful iterator that adheres to
+// FlatDatasetIterator interface. This allows the caller to iterate over every
+// element in the dataset, including elements nested inside sequences.
+//
+// Important note: if the Dataset changes during the iteration (e.g. if elements
+// are added or removed), those elements will not be included until a new
+// iterator is created.
+//
+// If you don't need to receive elements on a/ channel, and don't want to worry
+// about always exhausting this iterator, this is the best and safest way to
+// iterate over a Dataset.
+func (d *Dataset) FlatStatefulIterator() FlatDatasetIterator {
+	flatElems := flatElemsWrapper{}
+	flatSliceBuilder(d.Elements, &flatElems)
+	// fmt.Println("Dataset:", flatElems.flatElems)
+	iter := flatIterator{flattenedDataset: flatElems.flatElems}
+	return &iter
+}
+
+func flatSliceBuilder(datasetElems []*Element, flatElems *flatElemsWrapper) {
```

Comment:
I think you could probably change this signature to:

```go
func flatSliceBuilder(datasetElems []*Element, flatElems []*Element) []*Element {
     // add and return flat elements
}
```

... and remove the ```flatElemsWrapper``` type entirely. Right now you have to do a pointer-chase every time you append an object. It seems a little more streamlined to just pass and return the slice you want to append to rather than doing it as an in-place mutation using a pointer.

Reply:
+1 this is something I wanted to revisit in my draft before submitting, definitely agree.